### PR TITLE
Add SQL Big Data Cluster node for Aris clusters

### PR DESCRIPTION
**Pull request by t-rufang**

### CODE REVIEW

**Comment by t-rufang** [comment]
Close and reopen this PR to trigger vscjavaci build

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java [hunk 1]
```diff
@@ -218,74 +232,122 @@ AzureManager getAzureManager() {
                 .orElse(new ArrayList<>());
     }
 
-    public synchronized ImmutableList<IClusterDetail> getClusterDetails() {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java]: We can consider not changing this.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java [hunk 2]
```diff
@@ -218,74 +232,122 @@ AzureManager getAzureManager() {
...
     }
 
+    /**
+     * Get all cluster details except SQL Big Data clusters. This method will load all clusters first and then filter
+     * out SQL Big Data clusters
+     */
+    public synchronized ImmutableList<IClusterDetail> getClusterDetails() {
+        return ImmutableList.copyOf(
+                getAllClusterDetails().stream()
+                        .filter(clusterDetail -> !(clusterDetail instanceof SqlBigDataLivyLinkClusterDetail))
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java]: Can we move the filter to the method invoker?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java [hunk 3]
```diff
@@ -195,31 +208,50 @@ private boolean isBasicAuthSelected() {
...
+                    errorMessage = "The same name Livy Endpoint already exists in clusters";
+                } else if (ctrlProvider.doesClusterNameExistInAllHDInsightClusters(livyClusterNameField.getText())) {
+                    errorMessage = "Cluster Name already exists in clusters";
+                } else if (!StringUtils.isEmpty(yarnEndpointField.getText()) &&
+                        !ctrlProvider.isURLValid(yarnEndpointField.getText())) {
+                    errorMessage = "Yarn Endpoint is not a valid URL";
+                }
+                break;
+            case SQL_BIG_DATA_CLUSTER:
+                if (StringUtils.isBlank(livyEndpointField.getText()) ||
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: Are these checking the same as `LIVY_LINK_CLUSTER`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java [hunk 4]
```diff
@@ -185,6 +194,10 @@ private void createUIComponents() {
         yarnEndpointField = new HintTextField("(Optional)Example: http://hn0-spark2:8088");
     }
 
+    public SparkClusterType getSparkClusterType() {
+        return isHDInsightClusterSelected() ? SparkClusterType.HDINSIGHT_CLUSTER : SparkClusterType.LIVY_LINK_CLUSTER;
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: Only two types?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java (hunk 3 above)
Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: Not exactly the same. Cluster name duplicate validation and livy endpoint duplicate validation is different for HDI livy link cluster and SQL big data cluster.

---
In reply to: [237105022](https://github.com/Microsoft/azure-tools-for-java/pull/2235#discussion_r237105022) [](ancestors = 237105022)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java (hunk 4 above)
Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: We override `getSparkClusterType` method in `AddNewSparkClusterAction`. You can check the definition below:
```
            override fun getSparkClusterType(): SparkClusterType {
                return SparkClusterType.SQL_BIG_DATA_CLUSTER
            }

```

---
In reply to: [237105606](https://github.com/Microsoft/azure-tools-for-java/pull/2235#discussion_r237105606) [](ancestors = 237105606)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 5]
```diff
@@ -102,12 +103,22 @@ public SparkSubmissionJobUploadStorageWithUploadPathPanel getStorageWithUploadPa
     @NotNull
     protected ImmutableSortedSet<? extends IClusterDetail> getClusterDetails() {
         return ImmutableSortedSet.copyOf((x, y) -> x.getTitle().compareToIgnoreCase(y.getTitle()),
-                                         ClusterMetaDataService.getInstance().getCachedClusterDetails());
+                ClusterMetaDataService.getInstance().getCachedClusterDetails().stream()
+                        .filter(clusterDetail -> clusterDetail instanceof ClusterDetail ||
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: You can introduce a new filter method for this checking

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java [hunk 6]
```diff
@@ -52,30 +54,30 @@
     private JPanel clusterPanel;
     private JPanel authenticationPanel;
     private JPanel comboBoxPanel;
-    private JComboBox clusterComboBox;
-    private JPanel clusterCardsPanel;
+    protected JComboBox clusterComboBox;
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: Are these `protected` changes necessary?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java [hunk 7]
```diff
@@ -265,26 +273,39 @@ public synchronized  void addEmulatorCluster(EmulatorClusterDetail emulatorClust
...
-        hdinsightAdditionalClusterDetails.remove(hdInsightClusterDetail);
+    public synchronized void removeHDInsightAdditionalCluster(@NotNull IClusterDetail hdInsightClusterDetail) {
+        assert hdInsightClusterDetail instanceof HDInsightAdditionalClusterDetail ||
+                hdInsightClusterDetail instanceof HDInsightLivyLinkClusterDetail : "trying to remove a non-hdinsight-linked cluster";
+        additionalClusterDetails.remove(hdInsightClusterDetail);
         ClusterMetaDataService.getInstance().removeClusterFromCache(hdInsightClusterDetail);
+        saveAdditionalClusters();
+    }
 
+    public synchronized void addSqlBigDataAdditionalCluster(@NotNull IClusterDetail sqlBigDataClusterDetail) {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java]: Do we need these special methods for SqlBigDataAdditionalCluster?

Comment (t-rufang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java]: They are of no use, removed in the new commit.

---
In reply to: [237775209](https://github.com/Microsoft/azure-tools-for-java/pull/2235#discussion_r237775209) [](ancestors = 237775209)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java (hunk 6 above)
Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/serverexplore/ui/AddNewClusterForm.java]: Refer to #2410

---
In reply to: [237773385](https://github.com/Microsoft/azure-tools-for-java/pull/2235#discussion_r237773385) [](ancestors = 237773385)

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sqlbigdata/serverexplore/SqlBigDataClusterModule.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import com.microsoft.tooling.msservices.serviceexplorer.Node;
+import com.microsoft.tooling.msservices.serviceexplorer.NodeActionEvent;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SqlBigDataClusterModule extends HDInsightRootModule implements ILogger {
+    private static final String ARIS_SERVICE_MODULE_ID = SqlBigDataClusterModule.class.getName();
+    private static final String BASE_MODULE_NAME = "SQL Big Data Cluster";
+    private static final String ICON_PATH = CommonConst.HDExplorerIcon_16x_Path;
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sqlbigdata/serverexplore/SqlBigDataClusterModule.java]: We need a new icon?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/serverexplore/AddNewClusterCtrlProvider.java [hunk 9]
```diff
@@ -86,24 +85,30 @@ public boolean isURLValid(@NotNull String url) {
...
      * @param clusterName
-     * @return whether livy endpoint exists or not
+     * @return whether cluster name exists or not
      */
-    public boolean doesClusterNameExistInAllClusters(@NotNull String clusterName) {
-        return ClusterManagerEx.getInstance().getCachedClusters().stream().anyMatch(clusterDetail ->
-                clusterDetail.getName().equals(clusterName));
+    public boolean doesClusterNameExistInAllHDInsightClusters(@NotNull String clusterName) {
+        return ClusterManagerEx.getInstance().getCachedClusters().stream()
+                .filter(clusterDetail -> clusterDetail instanceof ClusterDetail ||
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/serverexplore/AddNewClusterCtrlProvider.java]: Can the filter reuse you just added?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/serverexplore/HDInsightRootModuleImpl.java [hunk 10]
```diff
@@ -71,7 +78,12 @@ public void refreshWithoutAsync() {
             removeAllChildNodes();
 
             // refresh the cluster list with invalidating the cache
-            clusterDetailList = ClusterManagerEx.getInstance().getClusterDetails();
+            clusterDetailList = ClusterManagerEx.getInstance().getClusterDetails().stream()
+                    .filter(clusterDetail -> clusterDetail instanceof ClusterDetail ||
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/serverexplore/HDInsightRootModuleImpl.java]: Reuse?

**Comment by t-rufang** [comment]
`vscjavaci` build failed
```
55c55
< |    |    |    \--- com.nimbusds:nimbus-jose-jwt:[5.5,) -> 6.4.1
---
> |    |    |    \--- com.nimbusds:nimbus-jose-jwt:[5.5,) -> 6.4.2
+ echo ALERT: New dependencies introduced!

```

**Pull request merged**